File: 01_two_sum/two_sum.cpp
#include <iostream>
#include <vector>
#include <map>

using namespace std;

class Solution{
public:
    // time limit exceeded
    vector<int> twoSum(vector<int> &numbers, int target) {
        vector<int> res;
        for (int i = 0; i < numbers.size() - 1; i++) {
            for (int j = i + 1; j < numbers.size() - 1; j++) {
                if (numbers[i] + numbers[j] == target) {
                    res.push_back(i + 1);
                    res.push_back(j + 1);
                }
            }
        }
        return res;
    }

    vector<int> twoSum2(vector<int> nums, int target) {
        vector<int> res;
        map<int, int> hmap;
        hmap.clear();
        for (int i = 0; i < nums.size(); i++) {
            hmap[nums[i]] = i;
        }
        
        for (map<int, int>::const_iterator it = hmap.begin(); it != hmap.end(); ++it) {
        cout << it->first << ' ' << it->second << endl;
        }
        
        cout << "-------";

        for (int i=0; i < nums.size(); i++) {
            if (hmap.find((target - nums[i])) != hmap.end()) {
                if (i < hmap[target - nums[i]]) {
                    res.push_back(i + 1);
                    res.push_back(hmap[target - nums[i]] + 1);
                    return res;
                }
                if (i > hmap[target - nums[i]]) {
                    res.push_back(hmap[target - nums[i]] + 1);
                    res.push_back(i + 1);
                    return res;
                }
            }
        }
    }
};

int main()
{
    vector<int> numbers;
    numbers.push_back(0);
    numbers.push_back(4);
    numbers.push_back(3);
    numbers.push_back(0);

    int target = 0;

    Solution so = Solution();
    // vector<int> res = so.twoSum(numbers, target);
    vector<int> res = so.twoSum2(numbers, target);
    for (int i = 0; i < res.size(); i++) {
        cout << res[i] << ' ';
    }
    cout << endl;
    return 0;
}

File: 01_two_sum/two_sum.py
class Solution(object):

    # time limit exceeded
    def twoSum(self, nums, target):
        ndict = dict()
        for index, num in enumerate(nums):
            ndict[num] = index
        for key, value in enumerate(nums):
            # time limit exceeded
            # if (target - value) in ndict.keys():
            # should take advantage of dict hasp map to find the key
            if ndict.has_key(target - value):
                if ndict[target - value] < key:
                    return ndict[target - value] + 1, key + 1
                if ndict[target - value] > key:
                    return key + 1, ndict[target -value] + 1

if __name__ == '__main__':
    nums = [3, 2, 4]
    target = 6
    so = Solution()
    print so.twoSum(nums, target)
